### ISSUE TITLE
Design by contract

### ISSUE THREAD

**Issue by privat** [privat/nit#1340]
It could be useful to define about a syntax to explicit contract for methods (pre and post conditions).
Such contracts can then be semantically checked by the frontend, asserted by engines (in debug mode) and exposed by the documentation.

Currently, some methods in libraries have free-style pre and post conditions.
Some specific nitunit can also be generalized as pre and post conditions.

The annotation syntax can be used to offer free style annotation

~~~nit
fun foo(a: Foo): Bar is
pre bla
pre do
  bla
  bla
end
post bla
post do
  bla
  bla
end
do
  bla
end
~~~

**Comment by R4PaSs** [comment]
I agree this could be interesting as it could clean the code, asserts could for the most part be replaced by pre and post conditions.

I also agree with the annotation example, this would allow for an easy implementation.

**Comment by Delja** [comment]
The pr #2790 provides a first implementation for the contract programming. Only the pre and post condition for the methods are available.

Exemple:

```nit
fun foo(a: Foo): Bar
is
	expects(a.int_attribut == 10) # Precondition
	ensures(self.int_attribut == 10, [other condition])# Postcondition
do
	# Body
end
```

The current syntax is different from the one proposed. The main reason is that the implementation of contracts uses the syntax of nit annotations that avoid the addition of keys words as well as a specific syntax in the grammar.

See #2790  for more explanation.

**Comment by Louis-Vincent** [comment]
To avoid writing multiple pre/post conditions you could assume that every whole boolean expression (not subexpression) is part of the contract, eg :

```nit
fun foo(x: Object): Object is
pre do
        self.x == 10 # one precondition
        var temp_result = [a computation] # not a precondition
        self.y == temp_result # a second precondition
end
post do
      self.x == 20 # one postcondition
end
do
end
```

Here a case of precondition that must be evaluated as a whole and not individually in the AST:
```nit
fun toto(...) is
pre do
        self.x == true # one precondition not two
end
do
end
```
Here because of `self.x == true` is a boolean expression AND `true` is an expression too, this should count as only one precondition.

I know, this code would throw a warning about unnecessary code since `self.x` already evaluates to a boolean and evaluating it against a literal boolean is useless.
